server/src/main/java/org/candlepin/policy/js/pool/PoolRules.java

```diff
@@ -244,15 +251,28 @@ private boolean hasBonusPool(List<Pool> pools) {
     /*
      * Returns null if invalid
      */
-    private String getVirtQuantity(String virtLimit, long quantity) {
+    private String getVirtQuantity(String virtLimit, long masterPoolQuantity, boolean hostLimited) {
+        if (hostLimited && ("unlimited".equals(virtLimit) || masterPoolQuantity == UNLIMITED_QUANTITY)) {
+            return String.valueOf(UNLIMITED_QUANTITY);
+        }
+
         if ("unlimited".equals(virtLimit)) {
             return virtLimit;
         }
 
         try {
             int virtLimitInt = Integer.parseInt(virtLimit);
+
             if (virtLimitInt > 0) {
-                return String.valueOf(virtLimitInt * quantity);
+                long quantity = virtLimitInt * masterPoolQuantity;
+
+                // Master pool quantity is negative (unlimited quantity)
+                // Set bonus pool quantity equal to virt_limit
+                if (quantity < 0) {
+                    return String.valueOf(virtLimitInt);
```

Comment:
This is now incorrect. If `quantity` is negative, that would mean that master pool is negative, so that means we want to return `"unlimited"`. Let's update this and the comment above it

Reply:
Since we treat unmapped_guest & Bonus pool same while calculating the quantities, this check is redundant.